File: code6804.pas
{$I STDINC.PAS}
{$IFDEF SEGATTRS}
{$C Moveable DemandLoad Discardable }
{$ENDIF}

	Unit Code6804;

Interface

        Uses NLS,
             AsmDef,AsmPars,AsmSub,CodePseu;


Implementation

TYPE
   BaseOrder=RECORD
	      Name:String[4];
	      Code:LongInt;
	     END;

CONST
   FixedOrderCnt=19;
   FixedOrders:ARRAY[1..FixedOrderCnt] OF BaseOrder=
	       ((Name:'CLRA'; Code:$00fbff),
		(Name:'CLRX'; Code:$b08000),
		(Name:'CLRY'; Code:$b08100),
		(Name:'COMA'; Code:$0000b4),
		(Name:'ROLA'; Code:$0000b5),
		(Name:'ASLA'; Code:$00faff),
		(Name:'INCA'; Code:$00feff),
		(Name:'INCX'; Code:$0000a8),
		(Name:'INCY'; Code:$0000a9),
		(Name:'DECA'; Code:$00ffff),
		(Name:'DECX'; Code:$0000b8),
		(Name:'DECY'; Code:$0000b9),
		(Name:'TAX' ; Code:$0000bc),
		(Name:'TAY' ; Code:$0000bd),
		(Name:'TXA' ; Code:$0000ac),
		(Name:'TYA' ; Code:$0000ad),
		(Name:'RTS' ; Code:$0000b3),
		(Name:'RTI' ; Code:$0000b2),
		(Name:'NOP' ; Code:$000020));

   RelOrderCnt=6;
   RelOrders:ARRAY[1..RelOrderCnt] OF BaseOrder=
	     ((Name:'BCC' ; Code:$40),
	      (Name:'BHS' ; Code:$40),
	      (Name:'BCS' ; Code:$60),
	      (Name:'BLO' ; Code:$60),
	      (Name:'BNE' ; Code:$00),
	      (Name:'BEQ' ; Code:$20));

   ALUOrderCnt=4;
   ALUOrders:ARRAY[1..ALUOrderCnt] OF BaseOrder=
	     ((Name:'ADD'; Code:$02),
	      (Name:'SUB'; Code:$03),
	      (Name:'CMP'; Code:$04),
	      (Name:'AND'; Code:$05));

   ModNone=-1;
   ModInd=0;    MModInd=1 SHL ModInd;
   ModDir=1;    MModDir=1 SHL ModDir;
   ModImm=2;    MModImm=1 SHL ModImm;

VAR
   AdrMode:ShortInt;
   AdrVal:Byte;

   CPU6804:CPUVar;


	PROCEDURE DecodeAdr(VAR Asc:String; MayImm:Boolean);
LABEL
   Found;
VAR
   OK:Boolean;

BEGIN
   AdrMode:=ModNone;

   IF NLS_StrCaseCmp(Asc,'(X)')=0 THEN
    BEGIN
     AdrMode:=ModInd; AdrVal:=$00; Goto Found;
    END;
   IF NLS_StrCaseCmp(Asc,'(Y)')=0 THEN
    BEGIN
     AdrMode:=ModInd; AdrVal:=$10; Goto Found;
    END;

   IF Asc[1]='#' THEN
    BEGIN
     AdrVal:=EvalIntExpression(Copy(Asc,2,Length(Asc)-1),Int8,OK);
     IF OK THEN AdrMode:=ModImm; Goto Found;
    END;

   AdrVal:=EvalIntExpression(Asc,Int8,OK);
   IF OK THEN
    BEGIN
     AdrMode:=ModDir; ChkSpace(SegData);
    END;

Found:
   IF (AdrMode=ModImm) AND (NOT MayImm) THEN
    BEGIN
     WrError(1350); AdrMode:=ModNone;
    END;
END;

	FUNCTION DecodePseudo:Boolean;
VAR
   z,z2:Integer;
   OK:Boolean;
   HVal8:ShortInt;
   HVal16:Integer;
BEGIN
   DecodePseudo:=True;

   IF Memo('SFR') THEN
    BEGIN
     CodeEquate(SegData,0,$ff);
     Exit;
    END;

   DecodePseudo:=False;
END;

	FUNCTION IsShort(Adr:Byte):Boolean;
BEGIN
   IsShort:=(Adr AND $fc)=$80;
END;

	PROCEDURE MakeCode_6804;
	Far;
VAR
   z,AdrInt:Integer;
   OK:Boolean;
BEGIN
   CodeLen:=0; DontPrint:=False;

   { zu ignorierendes }

   if Memo('') THEN Exit;

   { Pseudoanweisungen }

   IF DecodePseudo THEN Exit;

   IF DecodeMotoPseudo(True) THEN Exit;

   { Anweisungen ohne Argument }

   FOR z:=1 TO FixedOrderCnt DO
    WITH FixedOrders[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>0 THEN WrError(1110)
       ELSE
	BEGIN
	 IF Code SHR 16<>0 THEN CodeLen:=3
	 ELSE CodeLen:=1+Ord(Hi(Code)<>0);
	 IF CodeLen=3 THEN BAsmCode[0]:=Code SHR 16;
	 IF CodeLen>=2 THEN BAsmCode[CodeLen-2]:=Hi(Code);
	 BAsmCode[CodeLen-1]:=Lo(Code);
	END;
       Exit;
      END;

   { relative/absolute SprÅnge }

   FOR z:=1 TO RelOrderCnt DO
    WITH RelOrders[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE
	BEGIN
	 AdrInt:=EvalIntExpression(ArgStr[1],Int16,OK)-(EProgCounter+1);
	 IF OK THEN
	  IF (AdrInt<-16) OR (AdrInt>15) THEN WrError(1370)
	  ELSE
	   BEGIN
	    CodeLen:=1; BAsmCode[0]:=Code+(AdrInt AND $1f);
	    ChkSpace(SegCode);
	   END;
	END;
       Exit;
      END;

   IF (Memo('JSR')) OR (Memo('JMP')) THEN
    BEGIN
     IF ArgCnt<>1 THEN WrError(1110)
     ELSE
      BEGIN
       AdrInt:=EvalIntExpression(ArgStr[1],UInt12,OK);
       IF OK THEN
	BEGIN
	 CodeLen:=2; BAsmCode[1]:=Lo(AdrInt);
	 BAsmCode[0]:=$80+(Ord(Memo('JMP')) SHL 4)+(Hi(AdrInt) AND 15);
         ChkSpace(SegCode);
	END;
      END;
     Exit;
    END;

   { AKKU-Operationen }

   FOR z:=1 TO ALUOrderCnt DO
    WITH ALUOrders[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE
	BEGIN
	 DecodeAdr(ArgStr[1],True);
	 CASE AdrMode OF
	 ModInd:BEGIN
		 CodeLen:=1; BAsmCode[0]:=$e0+AdrVal+Code;
		END;
	 ModDir:BEGIN
		 CodeLen:=2; BAsmCode[0]:=$f8+Code; BAsmCode[1]:=AdrVal;
		END;
	 ModImm:BEGIN
		 CodeLen:=2; BAsmCode[0]:=$e8+Code; BAsmCode[1]:=AdrVal;
		END;
	 END;
	END;
       Exit;
      END;

   { Datentransfer }

   IF (Memo('LDA')) OR (Memo('STA')) THEN
    BEGIN
     IF ArgCnt<>1 THEN WrError(1110)
     ELSE
      BEGIN
       DecodeAdr(ArgStr[1],Memo('LDA'));
       AdrInt:=Ord(Memo('STA'));
       CASE AdrMode OF
       ModInd:BEGIN
	       CodeLen:=1; BAsmCode[0]:=$e0+AdrInt+AdrVal;
	      END;
       ModDir:IF IsShort(AdrVal) THEN
	       BEGIN
		CodeLen:=1; BAsmCode[0]:=$ac+(AdrInt SHL 4)+(AdrVal AND 3);
	       END
	      ELSE
	       BEGIN
		CodeLen:=2; BAsmCode[0]:=$f8+AdrInt; BAsmCode[1]:=AdrVal;
	       END;
       ModImm:BEGIN
	       CodeLen:=2; BAsmCode[0]:=$e8+AdrInt; BAsmCode[1]:=AdrVal;
	      END;
       END;
      END;
     Exit;
    END;

   IF (Memo('LDXI')) OR (Memo('LDYI')) THEN
    BEGIN
     IF ArgCnt<>1 THEN WrError(1110)
     ELSE IF ArgStr[1][1]<>'#' THEN WrError(1350)
     ELSE
      BEGIN
       BAsmCode[2]:=EvalIntExpression(Copy(ArgStr[1],2,Length(ArgStr[1])-1),Int8,OK);
       IF OK THEN
	BEGIN
	 CodeLen:=3; BAsmCode[0]:=$b0;
	 BAsmCode[1]:=$80+Ord(Memo('LDYI'));
	END;
      END;
     Exit;
    END;

   IF Memo('MVI') THEN
    BEGIN
     IF ArgCnt<>2 THEN WrError(1110)
     ELSE IF ArgStr[2][1]<>'#' THEN WrError(1350)
     ELSE
      BEGIN
       BAsmCode[1]:=EvalIntExpression(ArgStr[1],Int8,OK);
       IF OK THEN
	BEGIN
	 ChkSpace(SegData);
	 BAsmCode[2]:=EvalIntExpression(Copy(ArgStr[2],2,Length(ArgStr[2])-1),Int8,OK);
	 IF OK THEN
	  BEGIN
	   BAsmCode[0]:=$b0; CodeLen:=3;
	  END;
	END;
      END;
     Exit;
    END;

   { Read/Modify/Write-Operationen }

   IF (Memo('INC')) OR (Memo('DEC')) THEN
    BEGIN
     IF ArgCnt<>1 THEN WrError(1110)
     ELSE
      BEGIN
       DecodeAdr(ArgStr[1],False);
       AdrInt:=Ord(Memo('DEC'));
       CASE AdrMode OF
       ModInd:BEGIN
	       CodeLen:=1; BAsmCode[0]:=$e6+AdrInt+AdrVal;
	      END;
       ModDir:IF IsShort(AdrVal) THEN
	       BEGIN
		CodeLen:=1; BAsmCode[0]:=$a8+(AdrInt SHL 4)+(AdrVal AND 3);
	       END
	      ELSE
	       BEGIN
		CodeLen:=2; BAsmCode[0]:=$fe+AdrInt; BAsmCode[1]:=AdrVal;
	       END;
       END;
      END;
     Exit;
    END;

   { Bitbefehle }

   IF (Memo('BSET')) OR (Memo('BCLR')) THEN
    BEGIN
     IF ArgCnt<>2 THEN WrError(1110)
     ELSE
      BEGIN
       AdrVal:=EvalIntExpression(ArgStr[1],UInt3,OK);
       IF OK THEN
	BEGIN
	 BAsmCode[0]:=$d0+(Ord(Memo('BSET')) SHL 3)+AdrVal;
	 BAsmCode[1]:=EvalIntExpression(ArgStr[2],Int8,OK);
	 IF OK THEN
	  BEGIN
	   CodeLen:=2; ChkSpace(SegData);
	  END;
	END;
      END;
     Exit;
    END;

   IF (Memo('BRSET')) OR (Memo('BRCLR')) THEN
    BEGIN
     IF ArgCnt<>3 THEN WrError(1110)
     ELSE
      BEGIN
       AdrVal:=EvalIntExpression(ArgStr[1],UInt3,OK);
       IF OK THEN
	BEGIN
	 BAsmCode[0]:=$c0+(Ord(Memo('BRSET')) SHL 3)+AdrVal;
	 BAsmCode[1]:=EvalIntExpression(ArgStr[2],Int8,OK);
	 IF OK THEN
	  BEGIN
	   ChkSpace(SegData);
	   AdrInt:=EvalIntExpression(ArgStr[3],Int16,OK)-(EProgCounter+3);
	   IF OK THEN
	    IF (AdrInt<-128) OR (AdrInt>127) THEN WrError(1370)
	    ELSE
	     BEGIN
	      ChkSpace(SegCode); BAsmCode[2]:=AdrInt AND $ff; CodeLen:=3;
	     END;
	  END;
	END;
      END;
     Exit;
    END;

   WrXError(1200,OpPart);
END;

	FUNCTION ChkPC_6804:Boolean;
	Far;
BEGIN
   IF ProgCounter<0 THEN ChkPC_6804:=False
   ELSE
    CASE ActPC OF
    SegCode:ChkPC_6804:=ProgCounter<=$fff;
    SegData:ChkPC_6804:=ProgCounter<=$ff;
    ELSE ChkPC_6804:=False;
    END;
END;

	FUNCTION IsDef_6804:Boolean;
	Far;
BEGIN
   IsDef_6804:=Memo('SFR');
END;

        PROCEDURE SwitchFrom_6804;
        Far;
BEGIN
END;

	PROCEDURE SwitchTo_6804;
	Far;
BEGIN
   TurnWords:=False; ConstMode:=ConstModeMoto; SetIsOccupied:=False;

   PCSymbol:='PC'; HeaderID:=$64; NOPCode:=$20;
   DivideChars:=','; HasAttrs:=False;

   ValidSegs:=[SegCode,SegData];
   Grans[SegCode]:=1; ListGrans[SegCode]:=1; SegInits[SegCode]:=0;
   Grans[SegData]:=1; ListGrans[SegData]:=1; SegInits[SegData]:=0;

   MakeCode:=MakeCode_6804; ChkPC:=ChkPC_6804; IsDef:=IsDef_6804;
   SwitchFrom:=SwitchFrom_6804;
END;

BEGIN
   CPU6804:=AddCPU('6804',SwitchTo_6804);
END.

File: code6804.c
/* code6804.c */
/*****************************************************************************/
/* SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only                     */
/*                                                                           */
/* AS-Portierung                                                             */
/*                                                                           */
/* AS-Codegenerator Motorola/ST 6804                                         */
/*                                                                           */
/*****************************************************************************/

#include "stdinc.h"

#include <string.h>

#include "bpemu.h"
#include "strutil.h"
#include "asmdef.h"
#include "asmsub.h"
#include "asmpars.h"
#include "asmitree.h"
#include "codepseudo.h"
#include "motpseudo.h"
#include "codevars.h"
#include "errmsg.h"

#include "code6804.h"

typedef struct
{
  char *Name;
  LongInt Code;
} BaseOrder;

#define FixedOrderCnt 19

enum
{
  ModNone = -1,
  ModInd = 0,
  ModDir = 1,
  ModImm = 2
};

#define MModInd (1 << ModInd)
#define MModDir (1 << ModDir)
#define MModImm (1 << ModImm)

static ShortInt AdrMode;
static Byte AdrVal;

static CPUVar CPU6804;

static BaseOrder *FixedOrders;

/*--------------------------------------------------------------------------*/

static void DecodeAdr(const tStrComp *pArg, Boolean MayImm)
{
  Boolean OK;

  AdrMode = ModNone;

  if (!as_strcasecmp(pArg->Str, "(X)"))
  {
    AdrMode = ModInd;
    AdrVal = 0x00;
    goto chk;
  }
  if (!as_strcasecmp(pArg->Str, "(Y)"))
  {
    AdrMode = ModInd;
    AdrVal = 0x10;
    goto chk;
  }

  if (*pArg->Str == '#')
  {
    AdrVal = EvalStrIntExpressionOffs(pArg, 1, Int8, &OK);
    if (OK)
      AdrMode = ModImm;
    goto chk;
  }

  AdrVal = EvalStrIntExpression(pArg, Int8, &OK);
  if (OK)
    AdrMode = ModDir;

chk:
  if ((AdrMode == ModImm) && (!MayImm))
  {
    WrError(ErrNum_InvAddrMode); AdrMode = ModNone;
  }
}

static Boolean IsShort(Byte Adr)
{
  return ((Adr & 0xfc) == 0x80);
}

/*--------------------------------------------------------------------------*/

/* Anweisungen ohne Argument */

static void DecodeFixed(Word Index)
{
  const BaseOrder *pOrder = FixedOrders + Index;

  if (ChkArgCnt(0, 0))
  {
    if ((pOrder->Code >> 16) != 0)
      CodeLen = 3;
    else
      CodeLen = 1 + Ord(Hi(pOrder->Code) != 0);
    if (CodeLen == 3)
      BAsmCode[0] = pOrder->Code >> 16;
    if (CodeLen >= 2)
      BAsmCode[CodeLen - 2] = Hi(pOrder->Code);
    BAsmCode[CodeLen - 1] = Lo(pOrder->Code);
  }
}

/* relative/absolute Spruenge */

static void DecodeRel(Word Code)
{
  if (ChkArgCnt(1, 1))
  {
    tEvalResult EvalResult;
    Integer AdrInt = EvalStrIntExpressionWithResult(&ArgStr[1], Int16, &EvalResult) - (EProgCounter() + 1);

    if (EvalResult.OK)
    {
      if (!mSymbolQuestionable(EvalResult.Flags) && ((AdrInt < -16) || (AdrInt > 15))) WrError(ErrNum_JmpDistTooBig);
      else
      {
        CodeLen = 1;
        BAsmCode[0] = Code + (AdrInt & 0x1f);
        ChkSpace(SegCode, EvalResult.AddrSpaceMask);
      }
    }
  }
}

static void DecodeJSR_JMP(Word Code)
{
  if (ChkArgCnt(1, 1))
  {
    tEvalResult EvalResult;
    Word AdrInt = EvalStrIntExpressionWithResult(&ArgStr[1], UInt12, &EvalResult);

    if (EvalResult.OK)
    {
      CodeLen = 2;
      BAsmCode[1] = Lo(AdrInt);
      BAsmCode[0] = Code + (Hi(AdrInt) & 15);
      ChkSpace(SegCode, EvalResult.AddrSpaceMask);
    }
  }
}

/* AKKU-Operationen */

static void DecodeALU(Word Code)
{
  if (ChkArgCnt(1, 1))
  {
    DecodeAdr(&ArgStr[1], True);
    switch (AdrMode)
    {
      case ModInd:
        CodeLen = 1;
        BAsmCode[0] = 0xe0 + AdrVal + Code;
        break;
      case ModDir:
        CodeLen = 2;
        BAsmCode[0] = 0xf8 + Code;
        BAsmCode[1] = AdrVal;
        break;
      case ModImm:
        CodeLen = 2;
        BAsmCode[0] = 0xe8 + Code;
        BAsmCode[1] = AdrVal;
        break;
    }
  }
}

/* Datentransfer */

static void DecodeLDA_STA(Word Code)
{
  if (ChkArgCnt(1, 1))
  {
    DecodeAdr(&ArgStr[1], !Code);
    switch (AdrMode)
    {
      case ModInd:
        CodeLen = 1;
        BAsmCode[0] = 0xe0 + Code + AdrVal;
        break;
      case ModDir:
        if (IsShort(AdrVal))
        {
          CodeLen = 1;
          BAsmCode[0] = 0xac + (Code << 4) + (AdrVal & 3);
        }
        else
        {
          CodeLen = 2;
          BAsmCode[0] = 0xf8 + Code;
          BAsmCode[1] = AdrVal;
        }
        break;
      case ModImm:
        CodeLen = 2;
        BAsmCode[0] = 0xe8 + Code;
        BAsmCode[1] = AdrVal;
        break;
    }
  }
}

static void DecodeLDXI_LDYI(Word Code)
{
  if (!ChkArgCnt(1, 1));
  else if (*ArgStr[1].Str != '#') WrError(ErrNum_InvAddrMode);
  else
  {
    Boolean OK;

    BAsmCode[2] = EvalStrIntExpressionOffs(&ArgStr[1], 1, Int8, &OK);
    if (OK)
    {
      CodeLen = 3;
      BAsmCode[0] = 0xb0;
      BAsmCode[1] = Code;
    }
  }
}

static void DecodeMVI(Word Code)
{
  UNUSED(Code);

  if (!ChkArgCnt(2, 2));
  else if (*ArgStr[2].Str != '#') WrError(ErrNum_InvAddrMode);
  else
  {
    tEvalResult EvalResult;

    BAsmCode[1] = EvalStrIntExpressionWithResult(&ArgStr[1], Int8, &EvalResult);
    if (EvalResult.OK)
    {
      ChkSpace(SegData, EvalResult.AddrSpaceMask);
      BAsmCode[2] = EvalStrIntExpressionOffsWithResult(&ArgStr[2], 1, Int8, &EvalResult);
      if (EvalResult.OK)
      {
        BAsmCode[0] = 0xb0;
        CodeLen = 3;
      }
    }
  }
}

/* Read/Modify/Write-Operationen */

static void DecodeINC_DEC(Word Code)
{
  if (ChkArgCnt(1, 1))
  {
    DecodeAdr(&ArgStr[1], False);
    switch (AdrMode)
    {
      case ModInd:
        CodeLen = 1;
        BAsmCode[0] = 0xe6 + Code + AdrVal;
        break;
      case ModDir:
        if (IsShort(AdrVal))
        {
          CodeLen = 1;
          BAsmCode[0] = 0xa8 + (Code << 4) + (AdrVal & 3);
        }
        else
        {
          CodeLen = 2;
          BAsmCode[0] = 0xfe + Code;
          BAsmCode[1] = AdrVal;
        }
        break;
    }
  }
}

/* Bitbefehle */

static void DecodeBSET_BCLR(Word Code)
{
  if (ChkArgCnt(2, 2))
  {
    tEvalResult EvalResult;
    Byte Bit = EvalStrIntExpressionWithResult(&ArgStr[1], UInt3, &EvalResult);
    if (EvalResult.OK)
    {
      BAsmCode[0] = Code + Bit;
      BAsmCode[1] = EvalStrIntExpressionWithResult(&ArgStr[2], Int8, &EvalResult);
      if (EvalResult.OK)
      {
        CodeLen = 2;
        ChkSpace(SegData, EvalResult.AddrSpaceMask);
      }
    }
  }
}

static void DecodeBRSET_BRCLR(Word Code)
{
  if (ChkArgCnt(3, 3))
  {
    tEvalResult EvalResult;
    Byte Bit = EvalStrIntExpressionWithResult(&ArgStr[1], UInt3, &EvalResult);

    if (EvalResult.OK)
    {
      BAsmCode[0] = Code + Bit;
      BAsmCode[1] = EvalStrIntExpressionWithResult(&ArgStr[2], Int8, &EvalResult);
      if (EvalResult.OK)
      {
        Integer AdrInt;

        ChkSpace(SegData, EvalResult.AddrSpaceMask);
        AdrInt = EvalStrIntExpressionWithResult(&ArgStr[3], Int16, &EvalResult) - (EProgCounter() + 3);
        if (EvalResult.OK)
        {
          if (!mSymbolQuestionable(EvalResult.Flags) && ((AdrInt < -128) || (AdrInt > 127))) WrError(ErrNum_JmpDistTooBig);
          else
          {
            ChkSpace(SegCode, EvalResult.AddrSpaceMask);
            BAsmCode[2] = AdrInt & 0xff;
            CodeLen = 3;
          }
        }
      }
    }
  }
}

static void DecodeSFR(Word Code)
{
  UNUSED(Code);

  CodeEquate(SegData, 0, 0xff);
}

/*--------------------------------------------------------------------------*/

static void AddFixed(const char *NName, LongInt NCode)
{
  if (InstrZ >= FixedOrderCnt) exit(255);
  FixedOrders[InstrZ].Code = NCode;
  AddInstTable(InstTable, NName, InstrZ++, DecodeFixed);
}

static void AddRel(const char *NName, LongInt NCode)   
{
  AddInstTable(InstTable, NName, NCode, DecodeRel);
}

static void AddALU(const char *NName, LongInt NCode)   
{
  AddInstTable(InstTable, NName, NCode, DecodeALU);
}

static void InitFields(void)
{
  InstTable = CreateInstTable(203);
  AddInstTable(InstTable, "JMP", 0x90, DecodeJSR_JMP);
  AddInstTable(InstTable, "JSR", 0x80, DecodeJSR_JMP);
  AddInstTable(InstTable, "LDA", 0, DecodeLDA_STA);
  AddInstTable(InstTable, "STA", 1, DecodeLDA_STA);
  AddInstTable(InstTable, "LDXI", 0x80, DecodeLDXI_LDYI);
  AddInstTable(InstTable, "LDYI", 0x81, DecodeLDXI_LDYI);
  AddInstTable(InstTable, "MVI", 0, DecodeMVI);
  AddInstTable(InstTable, "INC", 0, DecodeINC_DEC);
  AddInstTable(InstTable, "DEC", 1, DecodeINC_DEC);
  AddInstTable(InstTable, "BSET", 0xd8, DecodeBSET_BCLR);
  AddInstTable(InstTable, "BCLR", 0xd0, DecodeBSET_BCLR);
  AddInstTable(InstTable, "BRSET", 0xc8, DecodeBRSET_BRCLR);
  AddInstTable(InstTable, "BRCLR", 0xc0, DecodeBRSET_BRCLR);
  AddInstTable(InstTable, "SFR", 0, DecodeSFR);

  FixedOrders = (BaseOrder *) malloc(sizeof(BaseOrder) * FixedOrderCnt); InstrZ = 0;
  AddFixed("CLRA", 0x00fbff);
  AddFixed("CLRX", 0xb08000);
  AddFixed("CLRY", 0xb08100);
  AddFixed("COMA", 0x0000b4);
  AddFixed("ROLA", 0x0000b5);
  AddFixed("ASLA", 0x00faff);
  AddFixed("INCA", 0x00feff);
  AddFixed("INCX", 0x0000a8);
  AddFixed("INCY", 0x0000a9);
  AddFixed("DECA", 0x00ffff);
  AddFixed("DECX", 0x0000b8);
  AddFixed("DECY", 0x0000b9);
  AddFixed("TAX" , 0x0000bc);
  AddFixed("TAY" , 0x0000bd);
  AddFixed("TXA" , 0x0000ac);
  AddFixed("TYA" , 0x0000ad);
  AddFixed("RTS" , 0x0000b3);
  AddFixed("RTI" , 0x0000b2);
  AddFixed("NOP" , 0x000020);

  AddRel("BCC", 0x40);
  AddRel("BHS", 0x40);
  AddRel("BCS", 0x60);
  AddRel("BLO", 0x60);
  AddRel("BNE", 0x00);
  AddRel("BEQ", 0x20);

  AddALU("ADD", 0x02);
  AddALU("SUB", 0x03);
  AddALU("CMP", 0x04);
  AddALU("AND", 0x05);

  AddInstTable(InstTable, "DB", 0, DecodeMotoBYT);
  AddInstTable(InstTable, "DW", 0, DecodeMotoADR);
  AddInstTable(InstTable, "DS", 0, DecodeMotoDFS);
}

static void DeinitFields(void)
{
  DestroyInstTable(InstTable);

  free(FixedOrders);
}

/*--------------------------------------------------------------------------*/

static void MakeCode_6804(void)
{
  CodeLen = 0;
  DontPrint = False;

  /* zu ignorierendes */

  if (Memo(""))
    return;

  /* Pseudoanweisungen */

  if (DecodeMotoPseudo(True))
    return;

  if (!LookupInstTable(InstTable, OpPart.Str))
    WrStrErrorPos(ErrNum_UnknownInstruction, &OpPart);
}

static Boolean IsDef_6804(void)
{
  return (Memo("SFR"));
}

static void SwitchFrom_6804(void)
{
  DeinitFields();
}

static void SwitchTo_6804(void)
{
  TurnWords = False;
  SetIntConstMode(eIntConstModeMoto);

  PCSymbol = "PC";
  HeaderID = 0x64;
  NOPCode = 0x20;
  DivideChars = ",";
  HasAttrs = False;

  ValidSegs = (1 << SegCode) | (1 << SegData);
  Grans[SegCode] = 1; ListGrans[SegCode] = 1; SegInits[SegCode] = 0;
  SegLimits[SegCode] = 0xfff;
  Grans[SegData] = 1; ListGrans[SegData] = 1; SegInits[SegData] = 0;
  SegLimits[SegData] = 0xff;

  MakeCode = MakeCode_6804;
  IsDef = IsDef_6804;
  SwitchFrom = SwitchFrom_6804;
  InitFields();
}

void code6804_init(void)
{
  CPU6804 = AddCPU("6804", SwitchTo_6804);
}
